pyquil/quil.py

```diff
@@ -587,19 +651,97 @@ def get_qubits(self, indices: bool = True) -> Set[QubitDesignator]:
         """
         qubits: Set[QubitDesignator] = set()
         for instr in self.instructions:
-            if isinstance(instr, (Gate, Measurement)):
+            if isinstance(
+                instr,
+                (
+                    Gate,
+                    Measurement,
+                    ResetQubit,
+                    Pulse,
+                    Capture,
+                    RawCapture,
+                    ShiftFrequency,
+                    SetFrequency,
+                    SetPhase,
+                    ShiftPhase,
+                    SwapPhase,
+                    SetScale,
+                ),
+            ):
                 qubits |= instr.get_qubits(indices=indices)
         return qubits
 
-    def is_protoquil(self) -> bool:
+    def match_calibrations(self, instr: AbstractInstruction) -> Optional[CalibrationMatch]:
+        """
+        Attempt to match a calibration to the provided instruction.
+
+        Note: preference is given to later calibrations, i.e. in a program with
+
+          DEFCAL X 0:
+              <a>
+
+          DEFCAL X 0:
+             <b>
+
+        the second calibration, with body <b>, would be the calibration matching `X 0`.
+
+        :param instr: An instruction.
+        :returns: a CalibrationMatch object, if one can be found.
+        """
+        if isinstance(instr, (Gate, Measurement)):
+            for cal in reversed(self.calibrations):
+                match = match_calibration(instr, cal)
+                if match is not None:
+                    return match
+
+        return None
+
+    def get_calibration(
+        self, instr: AbstractInstruction
+    ) -> Optional[Union[DefCalibration, DefMeasureCalibration]]:
+        """
+        Get the calibration corresponding to the provided instruction.
+
+        :param instr: An instruction.
+        :returns: A matching Quilt calibration definition, if one exists.
+        """
+        match = self.match_calibrations(instr)
+        if match:
+            return match.cal
+
+        return None
+
+    def calibrate(self, instr: AbstractInstruction) -> List[AbstractInstruction]:
+        """
+        Expand an instruction into its calibrated definition.
+
+        If a calibration definition matches the provided instruction, then the definition
+        body is returned with appropriate substitutions made for parameters and qubit
+        arguments. If no calibration definition matches, then the original instruction
+        is returned.
+
+        :param instr: An instruction.
+        :returns: A list of instructions, with the active calibration expanded.
+        """
+        # TODO: recursively expand?
```

Comment:
- [ ] TODO

Reply:
I'm not sure I understand the issue here. Isn't the body of a `DEFCAL` [a list of analog instructions](https://github.com/rigetti/quil/blob/master/rfcs/analog/spec_changes.md#defining-calibrations) which do not have associated calibrations?

Reply:
Following up from our chat, the calibration can contain other instructions which may themselves be calibrated